spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dsl/JobParser.java

```diff
@@ -0,0 +1,270 @@
+ * limitations under the License.
+ */
+
+package org.springframework.xd.dirt.job.dsl;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Parse a Batch DSL Job specification.
+ *
+ * @author Andy Clement
+ */
+public class JobParser {
+
+	private Tokens tokens;
+
+	public JobParser() {
+	}
+
+	/**
+	 * Parse a job flow definition into an abstract syntax tree (AST).
+	 *
+	 * @param the textual job specification
+	 * @return the AST for the parsed stream
+	 * @throws JobSpecificationException
+	 */
+	public JobSpecification parse(String jobSpecification) {
+		tokens = new Tokens(jobSpecification);
+		JobNode jobNode = parseJobNode();
+		JobSpecification js = new JobSpecification(jobSpecification, jobNode);
+		if (tokens.hasNext()) {
+			throw new JobSpecificationException(tokens.getExpression(), tokens.peek().startpos,
+					JobDSLMessage.UNEXPECTED_DATA_AFTER_JOBSPEC,
+					toString(tokens.next()));
+		}
+		return js;
+	}
+
+	private JobNode parseJobNode() {
+		// Handle (...)
+		if (tokens.maybeEat(TokenKind.OPEN_PAREN)) {
+			JobNode jn = parseJobNode();
+			tokens.eat(TokenKind.CLOSE_PAREN);
+			return jn;
+		}
+		// Handle a split < ... >
+		if (tokens.peek(TokenKind.SPLIT_OPEN)) {
+			JobNode jn = parseSplit();
+			// is the split part of a flow? "<..> || b"
+			return maybeParseFlow(jn);
+		}
+		JobDescriptor jd = parseJobDescriptor();
+		// Handle a potential flow "a || b"
+		return maybeParseFlow(jd);
+	}
+
+	private JobNode maybeParseFlow(JobNode firstJobNodeInFlow) {
+		if (tokens.peek(TokenKind.DOUBLE_PIPE)) {
+			List<JobNode> jobNodes = new ArrayList<>();
+			jobNodes.add(firstJobNodeInFlow);
+			while (tokens.maybeEat(TokenKind.DOUBLE_PIPE)) {
+				JobNode nextNode = parseJobNode();
+				// If nextNode is a Flow node, merge it with this one
+				if (nextNode instanceof Flow) {
+					jobNodes.addAll(nextNode.getSeries());
+				}
+				else {
+					jobNodes.add(nextNode);
+				}
+			}
+			return new Flow(jobNodes);
+		}
+		else {
+			return firstJobNodeInFlow;
+		}
+	}
+
+	// '<' jobs ['&' jobs]+ '>'
+	private JobNode parseSplit() {
+		List<JobNode> flows = new ArrayList<>();
+		tokens.eat(TokenKind.SPLIT_OPEN);
+		// '<' jobSequence  [ '&' jobSequence]* '>'
+		flows.add(parseJobNode());
+		while (tokens.maybeEat(TokenKind.AMPERSAND)) {
+			flows.add(parseJobNode());
+		}
+		tokens.eat(TokenKind.SPLIT_CLOSE);
+		return new Split(flows);
+	}
+
+	// JobDescriptor can be a simple JobReference (reference to a pre-existing definition) or
+	// an inlined JobDefinition.
+	//
+	// JobReference := identifier:jobname [stateTransitions]*
+	// JobDefinition := identifier:jobmodulename identifier:jobname ['--'optionname'='optionvalue]* [stateTransitions]*
+	private JobDescriptor parseJobDescriptor() {
+		JobDescriptor jd = null;
+		// Double identifiers indicates job definition
+		if (tokens.peek(0, TokenKind.IDENTIFIER)) {
+			if (tokens.peek(1, TokenKind.IDENTIFIER)) {
```

Comment:
We don't yet support labelling (with ':'), the colon mentioned in the comments above is merely my crude grammar description.  'identifier:jobname' is reminding me that we expect an identifier and that identifier should be treated as a jobname.  More work will need doing if we decide to support labelling.  I included those suffixes in the comments so that when I see 'identifier identifier' I know what I'm supposed to be treating those identifiers as.